### DISCUSSION TITLE
Why is model_validator(mode='after') better than field_validator ?

### DISCUSSION THREAD

**Discussion by JackPTwomey** [pydantic/pydantic#7974] [Question] [answered]
From the v2.4 docs on [Validators](https://docs.pydantic.dev/latest/concepts/validators/#field-validators):

> However, in most cases where you want to perform validation using multiple field values, it is better to use @model_validator which is discussed in the section below.

Is there a particular reason why `model_validator(mode='after')` is better? It seems a little more clunky to write

**Comment by sydney-runkle** [comment, marked as answer]
Good question! The text that comes before that statement in the docs is helpful for contextualizing this claim:
> If you want to access values from another field inside a @field_validator, this may be possible using ValidationInfo.data, which is a dict of field name to field value. Validation is done in the order fields are defined, so you have to be careful when using ValidationInfo.data to not access a field that has not yet been validated/populated — in the code above, for example, you would not be able to access info.data['id'] from within name_must_contain_space.

The idea here being that if you want to access a variety of field values in your custom validator, using a `@model_validator` with `mode='after'` ensures that you have access to all of your fields, whereas in the case of the `@field_validator` approach, you have to make sure not to access a field that has not yet been validated / populated.

Perhaps we should say it's ~better~ safer to use `@model_validator` in this case.
